Add map refresh

PunkPun opened the pull request:
Closes #19420

This is my first code contribution to ORA, so there's probably a lot of room for improvement

The way map refresh works:
`MapCache` has 2 instances of `MapDirectoryTracker` tracking both the custom and official maps folders for changes respectively. When they detect something, they analyse what and if they need to take action, then they put or edit appropriate actions in `mapActionQueue`. Whenever `MapCache` is queried for maps, it is updated from the records in `mapActionQueue`

I needed to queue actions because `FileSystemWatcher` is spamming quite a lot and those actions are often in an illogical order. F.e. when I add a file it calls `Changed`,  then `Created`. Another example is when I add / edit an uncompressed map, then it spams a ton of actions. Furthermore maps are never deleted from `MapCache`. So if we were to update `MapCache` at every `FileSystemWatcher` event it would likely clog up memory with very similar maps. The queue is designed to address all of this by updating `MapCache` only when it is required

Map editor no longer updates the cache, since it already happens when when MapDirectoryTracker kicks in

The game now selects the last edited map when entering map chooser. If in lobby the selected map is updated it, it will switch to the updated version. I'm not sure if this change is necessary, but it would be fairly convenient.

PunkPun commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -378,4 +401,118 @@ public void Dispose()
+			watcher.EnableRaisingEvents = true;
+		}
+
+		void AddActionSub(MapAction mapAction, string fullpath, string oldFullPath = "")
```
I'm really unsure how I should call these functions

PunkPun commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -378,4 +401,118 @@ public void Dispose()
+		{
+			// ignore hidden files
+			if (mapAction != MapAction.Delete && (new FileInfo(fullpath).Attributes & FileAttributes.Hidden) == FileAttributes.Hidden)
+				return;
```
I ignore hidden files here. But this method marks deleted files as hidden. With the way I handled it, hidden deleted files still go through. They don't cause any troubles though

abcdefg30 replied:
What is the reason for ignoring hidden files?

PunkPun replied:
Tbf no reason, the engine already deals with garbage files in map directories. I just didn't like to constantly see hidden osx and zipping files when debugging

PunkPun commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -378,4 +401,118 @@ public void Dispose()
+		readonly MapClassification classification;
+
+		enum MapAction { Add, Delete, Update }
+		readonly Dictionary<string, MapAction> mapActionQueue = new Dictionary<string, MapAction>();
```
I'm happy to change this queue mechanism to whatever is standard in ora engine. Just point me towards an example

RoosterDragon replied:
The data type can be whatever collection is most appropriate for your use case. Your current selection seems fine to me.

PunkPun replied:
Great

Punsho commented:
Fixed most of the bugs

PunkPun commented:
Fixed all bugs and crashes I could find. I changed ComputeUID to include map.png. ~~This change should be done in another pr, but I added it here temporarily as it is a very noticeable bug~~ #19891

pchote commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -265,7 +265,7 @@ public static string ComputeUID(IReadOnlyPackage package)
 			{
 				foreach (var filename in contents)
-					if (filename.EndsWith(".yaml") || filename.EndsWith(".bin") || filename.EndsWith(".lua"))
+					if (filename.EndsWith(".yaml") || filename.EndsWith(".bin") || filename.EndsWith(".lua") || filename == "map.png")
```
Noting the discussion from discord so that this PR won't be merged before a fix: this will break RC map queries unless we gate it on a map version upgrade (keeping format 11 maps using the existing UID calculation).

PunkPun replied:
Separated this change into its own commit

RoosterDragon replied:
For the purposes of this PR about dynamic map updates, shall we remove the change in `ComputeUID` for now so we can deal with that separately?

PunkPun replied:
sure I'll remove it. It is dealt with by #19891

PunkPun committed:
Make the delete button not call MapCache every frame

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/MapChooserLogic.cs:
```diff
@@ -98,7 +98,7 @@ public class MapChooserLogic : ChromeLogic
 
 			var deleteMapButton = widget.Get<ButtonWidget>("DELETE_MAP_BUTTON");
-			deleteMapButton.IsDisabled = () => modData.MapCache[selectedUid].Class != MapClassification.User;
+			deleteMapButton.IsDisabled = () => currentTab != MapClassification.User;
```
The delete button was querying `mapCache` every frame, defeating the whole purpose of the queue. 

All it wanted to know is if the map chooser tab was changed. As as `currentTab` is already updated on a switch, there's no need to query `mapChache`

PunkPun commented:
rebased

PunkPun commented:
Now when a user picks a map in map chooser that has been updated while user was in map chooser window, the game will act as if nothing was selected and exit. 

I also separated `MapDirectoryTracker` class into it's own file and rebased

RoosterDragon commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -48,12 +53,22 @@ public MapCache(ModData modData)
 			sheetBuilder = new SheetBuilder(SheetType.BGRA);
 		}
 
+		public void UpdateMaps()
```
This can be private, we don't use an access keyword for methods using the default access so just `void UpdateMaps()` is fine.

PunkPun replied:
ok

RoosterDragon commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -48,12 +53,22 @@ public MapCache(ModData modData)
+			foreach (var tracker in mapDirectoryTrackers)
+			{
+				tracker.UpdateMaps();
+			}
```
For things on a single line, our style says to leave off the braces.

PunkPun replied:
ok

RoosterDragon commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -85,36 +100,44 @@ public void LoadMaps()
-								continue;
+					LoadMap(map, kv.Key, kv.Value, mapGrid);
+				}
+			}
```
Similarly the braces around the `LoadMap` call can go since it's on one line.

PunkPun replied:
ok

RoosterDragon replied:
Can also remove the next level of outer braces, so:
```c#
foreach (...)
    foreach (...)
        LoadMap(...);
```

RoosterDragon commented on OpenRA.Game/Map/MapDirectoryTracker.cs:
```diff
@@ -0,0 +1,121 @@
+{
+	public class MapDirectoryTracker
+	{
+		readonly FileSystemWatcher watcher;
```
Because `FileSystemWatcher ` is `IDisposable`, we need to ensure that logic is plumbed through, so we don't have file watchers living forever once we stop caring about them.

- Change the class to `public sealed class MapDirectoryTracker : IDisposable`
- Implement `public void Dispose` with a body of `watcher.Dispose();`
- In `MapCache`, update the `Dispose` method to also loop over and dispose every tracker.

PunkPun replied:
done

RoosterDragon commented on OpenRA.Game/Map/MapDirectoryTracker.cs:
```diff
@@ -0,0 +1,121 @@
+
+		void AddMapAction(MapAction mapAction, string fullpath, string oldFullPath = null)
+		{
+			dirty = true;
```
The `FileSystemWatcher` is going to raise these callbacks on a different thread, we'll want to lock to avoid threading issues.

Around the entire body of `AddMapAction` and `UpdateMaps` I would add `lock (mapActionQueue) { ... }`.

PunkPun replied:
I'm not entirely sure how `lock` works. When I try applying the lock statement throws this error `The name 'mapActionQueue' does not exist in the current context.`. It appears that I also need to include` RemoveSubDirs` in the lock statement

RoosterDragon replied:
e.g.

```c#
void AddMapAction(MapAction mapAction, string fullpath, string oldFullPath = null)
{
    lock (mapActionQueue)
    {
        // body of the AddMapAction method
    }
}
```

PunkPun replied:
done

RoosterDragon commented on OpenRA.Game/Map/MapDirectoryTracker.cs:
```diff
@@ -0,0 +1,121 @@
+			}
+		}
+
+		public void UpdateMaps()
```
Provide `MapCache` as a parameter here so we can avoid reaching into global state on `Game` which is never fun.

PunkPun replied:
done

RoosterDragon commented on OpenRA.Game/Map/MapDirectoryTracker.cs:
```diff
@@ -0,0 +1,121 @@
+			foreach (var mapAction in mapActionQueue)
+			{
+				var map = Game.ModData.MapCache.FirstOrDefault(x => x.Package?.Name == mapAction.Key && x.Status == MapStatus.Available);
+				if (map != null)
```
Because the expressions in this if/else are not one line, please add some braces.

PunkPun replied:
ok

RoosterDragon commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -39,6 +39,11 @@ public sealed class MapCache : IEnumerable<MapPreview>, IDisposable
+		readonly List<MapDirectoryTracker> mapDirectoryTrackers = new List<MapDirectoryTracker>();
+		public string LastUpdatedUid;
+
+		public Action<string, string> MapUpdated = (x, y) => { };
```
This should be an event handler: `public event Action<string, string> MapUpdated;`

Subscribers can do `MapUpdated += (...) => { ... };` to attach a handler to it.

The `MapCache` class is the only class that can invoke this by calling `MapUpdated(...)`. You won't be able to invoke this directly from the tracker. This is fine, because it means we're clearing up the ownership. I would recommend updating `LoadMap` to raise this event when it realises it is updating an existing entry. I would recommend adding a xmldoc comment to explain what the two parameters are for as well.

PunkPun replied:
I though it would be best to just keep it as action as I'm only implementing it in lobby. But I guess it wouldn't hurt to make it an event

PunkPun replied:
done

PunkPun replied:
I don't think `LoadMap` needs to raise this event, unless we want to raise it also when a map is added. Then we might as well call it when a map is removed

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -182,20 +194,30 @@ void ConnectionStateChanged(OrderManager om, string password, NetworkConnection
 					{
-						// Don't select the same map again
-						if (uid == map.Uid)
+						// Don't select the same map again, and handle map becoming unavailabble
```
Nit: Typo.

PunkPun replied:
fixed

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/MapChooserLogic.cs:
```diff
@@ -126,6 +126,18 @@ public class MapChooserLogic : ChromeLogic
+					SwitchTab(currentTab, itemTemplate);
+					return;
+				}
+			}
```
Having a `LastUpdatedUid` feels really inelegant. Can you explain to me what this logic is trying to achieve?

PunkPun replied:
I agree, it was even pretty buggy. I moved it to `LobbyLogic` 
It now relies on `MapUpdated` event, but inclusion of it made me make `UpdateMaps` public again. I didn't find another non-hacky solution to trigger an update

PunkPun replied:
Oh and for the explanation. The goal of it is when a user first enters map chooser, the last updated map is preselected. In basically all scenarios where a user updates or adds a map he wants to immediately chose it in the lobby. So this is a shortcut. 
Though it doesn't work if a user is outside the game lobby as `LobbyLogic` won't be subscribed to updates. I recon it is fine

RoosterDragon replied:
`UpdateMaps` being public is fine, it seems reasonable to ask a cache to refresh itself.

Much prefer the new logic flow you've created here, it's much nicer that the `LobbyLogic` is tracking the last updated ID for it's own purposes rather than relying on another class to do it.

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -129,6 +131,16 @@ void ConnectionStateChanged(OrderManager om, string password, NetworkConnection
 			Game.LobbyInfoChanged += UpdateSpawnOccupants;
 			Game.BeforeGameStart += OnGameStart;
 			Game.ConnectionStateChanged += ConnectionStateChanged;
+			modData.MapCache.MapUpdated = (string oldUid, string newUid) =>
```
When `MapUpdated` becomes an `event`, we're going to want to unsubscribe from the event handler when we're done.

Store a reference to the method in a field, and subscribe with `modData.MapCache.MapUpdated += trackMapUids;`

In the overridden `Dispose` method, also call `modData.MapCache.MapUpdated -= trackMapUids;`

PunkPun replied:
done

RoosterDragon commented on OpenRA.Game/Map/MapDirectoryTracker.cs:
```diff
@@ -0,0 +1,135 @@
+
+							// Provide game lobby with information about our latest update
+							if (lastUid != null && map.Uid != lastUid)
+								mapcache.CallMapUpdated(map.Uid, lastUid);
```
Whilst I'm happy if we only fire `MapUpdated` for updates and not add/deletes, the responsibility should still be on `MapCache` to choose when that happens rather than allowing any caller to trigger it via `CallMapUpdated`.

Could `LoadMap` have some logic to check if it's updating an existing entry rather than adding a new one, and fire the event?

PunkPun replied:
I moved `MapUpdated` to `LoadMap`. It is now also called when a map was added, letting me move `lastUpdatedUid` to `LobbyLogic`

RoosterDragon commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -39,6 +39,15 @@ public sealed class MapCache : IEnumerable<MapPreview>, IDisposable
+		/// Called when a map has been updated. Provides old and new uid's of the map
+		/// Not called when map is added or deleted
+		/// </summary>
+		public event Action<string, string> MapUpdated = (oldUid, newuid) => { };
```
Nitpicking: Use the casing of `UID` in the comment, and `newUid` in the parameter.

PunkPun replied:
fixed

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -182,20 +186,35 @@ void ConnectionStateChanged(OrderManager om, string password, NetworkConnection
+					});
+
+					// We make sure lastUpdatedUid is up to date
+					modData.MapCache.UpdateMaps();
```
If I'm following the logic here now, we're preselecting the map you've just changed on disk for your convenience? That seems handy.

I think a comment like `// Check for updated maps, if the user has edited a map we'll preselect it for them.` would be more descriptive as it's giving a 'why' instead of a 'what'.

PunkPun replied:
I'll change the comment right away

PunkPun commented:
Now when you force start or click on a ready checkbox, the game properly handles the selected map becoming unavailable. The previous behaviour was quite unpolished. On game start the screen become black, and then the lobby is refreshed.

PunkPun commented:
Handled a crash. Now if you try to play a map in `MissionSelector` that has been marked as unavailable you will exit `MissionSelector`. Same behaviour as Map Editor `MapChooser`

PunkPun committed:
Add dynamic map refresh

PunkPun commented:
> Now when you force start or click on a ready checkbox

I undid the checkbox behaviour (but kept force start). My solution felt a bit hacky and honestly unnecessary. It's a minor polish feature that can be added in a followup pr

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs:
```diff
@@ -190,7 +190,7 @@ void CreateMissionGroup(string title, IEnumerable<MapPreview> previews)
 					() => selectedMap != null && selectedMap.Uid == preview.Uid,
 					() => SelectMap(preview),
-					StartMissionClicked);
+					new Action(() => StartMissionClicked(onExit)));
```
Could this be just `() => StartMissionClicked(onExit)`?

PunkPun replied:
Yes. 
For some reason that didn't work for me at first, but now it seems correct

PunkPun commented:
rebased and changed console reporting from "Added 'map'" to "'map' was added"

I'm unsure wether we should keep this reporting. For me it made sense to exist, however when looking back on it, you wouldn't really use it for anything

Mailaender commented:
I am fine with it, as the map editor also reports file system changes that way.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6dba4500ac30c941a18d1e64b556a665062dbf51...c6c38f1ce85546d523ce716f0b8edeb6c198b2b7)
